### PR TITLE
write waterbody results to LAKEOUT files

### PR DESCRIPTION
**Pull request by shorvath-noaa**
Added the capability to write LAKEOUT netcdf files containing inflow, outflow, and depth of waterbodies. One file is created per time step of the run.

## Additions

- lakeout_output parameter to output parameters in yaml file (documentation also added to v3_doc.yaml)
- added lakeout directory in test directory where LAKEOUT netcdf files will be saved.
- write_waterbody_netcdf function to nhd_io.py file. This function writes one netcdf file per time step containing information for all waterbodies
- waterbodies_df and waterbody_types_df added as input parameters to nwm_output_generator function in __main__.py
- waterbodies_df and waterbody_types_df added as input parameters to nwm_output_generator function in output.py
- added wbdyo from lakeout_ouput parameter in nwm_output_generator function
- added creation of wbdy_df containing waterbody parameters. Added conditionals to only perform these calculations if waterbodies are being used.
- added upstream_array in mc_reach.pyx which is an empty array that is populated with the inflow for waterbodies. returned as 6th index

## Removals

-

## Changes

-

## Testing

1. Verified that netcdf files print to correct directory
2. Verified that no files are created if data assimilation is turned off, or if reaches are not separated by waterbodies
3. Verified that netcdf files created by write_waterbody_netcdf function contain all necessary components/metadata

## Notes

-

## Todos

-

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

**Comment by shorvath-noaa** [comment]
@awlostowski-noaa

**Comment by awlostowski-noaa** [comment]
please add a `.gitkeep` file in a new subdirectory `t-route/test/LowerColorado_TX/lakeout`. This placeholder allows the directory to exist in the repository such that when a user runs `test_AnA.yaml` there is a writing destination for lakeout files.

**Comment by awlostowski-noaa** [comment]
Able to run the test case successfully, after creating a `/lakeout` subdirectory. However, one concern is the time it takes to write data to `LAKEOUT` files. 

```
2022-04-19 05:12:33,203   DEBUG [output.py:213 - nwm_output_generator()]: writing LAKEOUT files took a total time of 10.108326196670532 seconds.
```

This is twice as long as the time required to do the actual routing calculation. I wonder where the bottle neck is?

File: src/troute-nwm/src/nwm_routing/output.py [hunk 1]
```diff
@@ -162,6 +192,26 @@ def nwm_output_generator(
     if test:
         flowveldepth.to_pickle(Path(test))
     
+    if wbdyo and not waterbodies_df.empty:
+        
+        LOG.info("- writing t-route flow results to LAKEOUT files")
+        start = time.time()
+        
+        for time_index, i in enumerate(wbdy_df.time.unique()):
```
Comment (awlostowski-noaa) [inline src/troute-nwm/src/nwm_routing/output.py:200]: Please write-out LAKEOUT files on coarser time intervals. Here, we are writing them for every 5-minute timestep, however we only need to write files at every `dt * qts_subdivisions` seconds (both of those parameters are entered by the user in the .yaml file under `compute_parameters[forcing_parameters]`). In the case of our testing case, `qts_subdivisions = 12` which indicates that there are 12 timesteps per WRF reporting interval. In other words WRF reports every 60 minutes (300 sec * 12 = 3600 sec = 1 hour), and so this is the frequency we expect to get (and write) data from/to CHRTOUT, and in this case, to LAKEOUT. 

The sheer amount of LAKEOUT files we are writing is one bottleneck.

**Review by awlostowski-noaa — CHANGES_REQUESTED** [review]
- file frequency
- create `/lakeout` subdirectory

**Comment by shorvath-noaa** [comment]
@awlostowski-noaa
- added the .gitkeep file in lakeout directory
- edited output.py to filter wbdy_df to only the needed timesteps according to dt*qts_subdivisions
- edited nhd_io.py to accommodate changes to wbdy_df

**Review by awlostowski-noaa — APPROVED** [review]
Just noting that the LAKEOUT files created by t-route assume that `feature_id` is an int64 variable type. Adam will check in with NCAR to make sure this isn't an issue.

File: src/troute-network/troute/nhd_io.py [hunk 2]
```diff
@@ -1559,3 +1559,274 @@ def lastobs_df_output(
...
+    
+    # Merge waterbodies_df with wbdy_data
+    full_wbdy_data = pd.merge(waterbodies_df,wbdy_data,on = 'ID')
+    
+    if wbdy_filepath:
+        
+        # open netCDF4 Dataset in write mode
+        with netCDF4.Dataset(
+            filename = wbdy_filepath + str(wbdy_time[0].strftime('%Y%m%d%H%M')) + '.LAKEOUT.nc',
+            mode = 'w',
```
Comment (awlostowski-noaa) [inline src/troute-network/troute/nhd_io.py:1612]: not sure if it is a permissions issue, but if I re-run the code before I delete the batch of LAKEOUT files previously created, then I get this error. 

```
PermissionError: [Errno 13] Permission denied: b'lakeout/202108231300.LAKEOUT.nc'
```
We need to figure out how to over-write existing files, somehow. I know there are numerous `mode` argument options - does one of them meet our needs? If not, we might need to delete existing files before we go about re-creating them (same outcome, as far as I can tell).

**Review by awlostowski-noaa — CHANGES_REQUESTED** [review]
Found one remaining issue related to compatibility with diffusive hybrid. Figured out how to deal with one of two remaining issues.

File: src/troute-network/troute/nhd_io.py (hunk 2 above)
Comment (shorvath-noaa) [reply src/troute-network/troute/nhd_io.py:1612]: Mode 'w' is write and is supposed to overwrite the file if it already exists. I am not seeing this PermissionError when I re-run the code without deleting the existing LAKEOUT files, so perhaps it is a permission issue.

But we could try mode 'a' (append) and simply re-write all of the information to an existing file?

Comment (shorvath-noaa) [reply src/troute-network/troute/nhd_io.py:1612]: Mode 'a' should also create a new file if one does not already exist.

Comment (awlostowski-noaa) [reply src/troute-network/troute/nhd_io.py:1612]: hm - I am no longer seeing this error either. I might have had a file open in another python instance when I re-ran the code.  Let's stick with `w`

**Comment by awlostowski-noaa** [comment]
Also, can you please replace lines 340 - 369 in `output.py` with the following snippet?
```python
    # Write out LastObs as netcdf.
    # This is only needed if 1) streamflow nudging is ON and 2) a lastobs output
    # folder is provided by the user.
    streamflow_da = data_assimilation_parameters.get('streamflow_da', None)
    nudging_true = streamflow_da.get('streamflow_nudging', None)
    if streamflow_da:
        lastobs_output_folder = streamflow_da.get(
            "lastobs_output_folder", None
            )

    if nudging_true and lastobs_output_folder:
        
        LOG.info("- writing lastobs files")
        start = time.time()
        
        nhd_io.lastobs_df_output(
            lastobs_df,
            dt,
            nts,
            t0,
            link_gage_df['gages'],
            lastobs_output_folder,
        )
        
        LOG.debug("writing lastobs files took %s seconds." % (time.time() - start))
```
I got an error when I had `streamflow_nudging` set to false, yet still had a valid directory entered for `lastobs_output_folder`. The changes proposed above make is such that "lastobs" files are only written if streamflow nudging is turned ON and there is a valid destination directory path provided by the user. 

I fully realize this is outside the scope of the pull request, but I'd like to go ahead and get this fix in while it is fresh in my mind.

File: src/troute-routing/troute/routing/compute.py [hunk 3]
```diff
@@ -1173,6 +1173,7 @@ def compute_diffusive_routing(
                 # place-holder for reservoir DA paramters
                 (np.asarray([]), np.asarray([]), np.asarray([]), np.asarray([]), np.asarray([])),
                 (np.asarray([]), np.asarray([]), np.asarray([]), np.asarray([]), np.asarray([])),
+                np.asarray([])
```
[the remark this answers is not in the document]
Comment (awlostowski-noaa) [reply src/troute-routing/troute/routing/compute.py:1176]: @shorvath-noaa Did we take care of this one?

[the remark this answers is not in the document]
Comment (shorvath-noaa) [reply src/troute-routing/troute/routing/compute.py]: Ok, taken care of.

**Pull request merged**